File: content/blog/search_suggestions_system/go/solution_trie.go
package search_suggestions_system

import (
	"sort"
)

func suggestedProductsTrie(products []string, searchWord string) [][]string {
	root := &trie{
		children: make(map[rune]*trie),
	}

	sort.Strings(products)
	for _, p := range products {
		root.Insert(p)
	}

	runeSearch := []rune(searchWord)
	worldStart, exist := root.children[runeSearch[0]]
	if !exist {
		return root.gd(runeSearch)
	}

	suggestions := worldStart.Closest(runeSearch[1:], 3)
	res := make([][]string, len(suggestions))
	prev := []string{}
	i := len(suggestions) - 1
	if i > 0 {
		prev = suggestions[i][:]
		res[i] = prev
		i--
	}

	for ; i >= 0; i-- {
		newItem := suggestions[i][:]
		if len(prev) == 0 {
			res[i] = newItem
			prev = newItem[:]
			continue
		} else if len(newItem) == 0 {
			res[i] = prev[:]
			continue
		}

		newItem = append(newItem, prev...)
		sort.Strings(newItem)
		if len(newItem) > 3 {
			newItem = newItem[:3]
		}

		res[i] = newItem
		prev = newItem[:]
	}

	return res
}

type trie struct {
	world    string
	sort     []rune
	children map[rune]*trie
}

func (t *trie) Insert(word string) {
	t.insert(word, []rune(word))
}

func (t *trie) insert(origin string, word []rune) {
	if len(word) == 0 {
		t.world = origin
		return
	}

	child, exist := t.children[word[0]]
	if !exist {
		child = &trie{
			children: make(map[rune]*trie),
			sort:     make([]rune, 0, 5),
		}

		t.children[word[0]] = child
		t.sort = append(t.sort, word[0])
	}

	child.insert(origin, word[1:])
}

func (t *trie) Closest(prefix []rune, count int) [][]string {
	res := make([][]string, 0, len(prefix))
	var excludeRune *rune
	if len(prefix) > 0 {
		excludeRune = &prefix[0]
	}

	res = append(res, t.exactWorlds(excludeRune, count))

	if len(prefix) == 0 {
		return res
	}

	child, exist := t.children[prefix[0]]
	if !exist {
		res = [][]string{t.exactWorlds(nil, count)}
		return append(res, t.gd(prefix)...)
	}

	return append(res, child.Closest(prefix[1:], count)...)
}

func (t *trie) exactWorlds(excludeRune *rune, n int) []string {
	res := make([]string, 0, n)
	if len(t.world) > 0 {
		res = append(res, t.world)
		n--
	}

	if n == 0 {
		return res
	}

	for _, childRune := range t.sort {
		if excludeRune != nil && childRune == *excludeRune {
			continue
		}
		childWorlds := t.children[childRune].exactWorlds(nil, n)
		if len(childWorlds) >= n {
			return append(res, childWorlds[:n]...)
		}

		res = append(res, childWorlds...)
		n = n - len(childWorlds)
		if n == 0 {
			break
		}
	}

	return res
}

func (t *trie) gd(prefix []rune) [][]string {
	res := make([][]string, 0, len(prefix))
	for range prefix {
		res = append(res, []string{})
	}

	return res
}

File: content/blog/search_suggestions_system/ts/solution_trie.ts
class Trie {
	word: string = '';
	sort: string[] = [];
	children: Map<string, Trie> = new Map();

	insert(word: string): void {
		this._insert(word, [...word]);
	}

	private _insert(origin: string, word: string[]): void {
		if (word.length === 0) {
			this.word = origin;
			return;
		}

		const char = word[0];
		let child = this.children.get(char);

		if (!child) {
			child = new Trie();
			this.children.set(char, child);
			this.sort.push(char);
		}

		child._insert(origin, word.slice(1));
	}

	closest(prefix: string[], count: number): string[][] {
		const res: string[][] = [];
		let excludeChar: string | null = null;

		if (prefix.length > 0) {
			excludeChar = prefix[0];
		}

		res.push(this.exactWords(excludeChar, count));

		if (prefix.length === 0) {
			return res;
		}

		const child = this.children.get(prefix[0]);
		if (!child) {
			res[0] = this.exactWords(null, count);
			return res.concat(this.gd(prefix));
		}

		return res.concat(child.closest(prefix.slice(1), count));
	}

	private exactWords(excludeChar: string | null, n: number): string[] {
		let res: string[] = [];
		if (this.word.length > 0) {
			res.push(this.word);
			n--;
		}

		if (n === 0) {
			return res;
		}

		for (const childChar of this.sort) {
			if (excludeChar !== null && childChar === excludeChar) {
				continue;
			}

			const childWords = this.children.get(childChar)!.exactWords(null, n);
			if (childWords.length >= n) {
				return res.concat(childWords.slice(0, n));
			}

			res = res.concat(childWords);
			n -= childWords.length;

			if (n === 0) {
				break;
			}
		}

		return res;
	}

	gd(prefix: string[]): string[][] {
		const res: string[][] = [];
		for (let i = 0; i < prefix.length; i++) {
			res.push([]);
		}
		return res;
	}
}

export function suggestedProductsTrie(products: string[], searchWord: string): string[][] {
	const root = new Trie();

	products.sort();

	for (const p of products) {
		root.insert(p);
	}

	const runeSearch = [...searchWord];
	const firstChar = runeSearch[0];
	const worldStart = root.children.get(firstChar);

	if (!worldStart) {
		return root.gd(runeSearch);
	}

	const suggestions = worldStart.closest(runeSearch.slice(1), 3);
	const res: string[][] = new Array(suggestions.length);
	let prev: string[] = [];
	let i = suggestions.length - 1;

	if (i >= 0) {
		prev = suggestions[i].slice();
		res[i] = prev;
		i--;
	}

	for (; i >= 0; i--) {
		let newItem = suggestions[i].slice();

		if (prev.length === 0) {
			res[i] = newItem;
			prev = newItem.slice();
			continue;
		} else if (newItem.length === 0) {
			res[i] = prev.slice();
			continue;
		}

		newItem = newItem.concat(prev);
		newItem.sort();

		if (newItem.length > 3) {
			newItem = newItem.slice(0, 3);
		}

		res[i] = newItem;
		prev = newItem.slice();
	}

	return res;
}
